pkg/app/route_wasm_test.go

```diff
@@ -0,0 +1,186 @@
+	DOM_ERROR = "_DOM_ERROR_"
+)
+
+// testComp is a very simple test component that renders as a text node containing its name
+type testComp struct {
+	Compo
+	name       string
+	hitCounter int
+}
+
+func (c *testComp) Render() UI {
+	c.hitCounter++
+	return Div().Body(Text(c.name))
+}
+
+// TestRoute tests the basic routing tables
+func TestRoute(t *testing.T) {
+
+	resetRoutes()
+	Route("/a", &testComp{name: "a"})
+	Route("/abc", &testComp{name: "abc"})
+	Route("/azz", &testComp{name: "azz"})
+	Route("/xyz", &testComp{name: "xyz"})
+
+	navExpect(t, "/abc", "abc")
+	navExpect(t, "/a", "a")
+	navExpect(t, "/x", NOT_FOUND)
+	navExpect(t, "", NOT_FOUND)
+	navExpect(t, "/abc?foo=bar", "abc") // query is not used in route lookup
+}
+
+// TestRouteWithRegexp tests some basic regex patterns and priority of Route vs RouteWithRegexp
+func TestRouteWithRegexp(t *testing.T) {
+
+	resetRoutes()
+	Route("/a", &testComp{name: "a"})
+	Route("/abc", &testComp{name: "abc"})
+	Route("/123", &testComp{name: "123"})
+
+	RouteWithRegexp("/x.*", &testComp{name: "x-star"})
+	RouteWithRegexp("/a.*", &testComp{name: "a-star"}) // test match when not first regex route
+
+	navExpect(t, "/a", "a") // specific routes should be found before regex
+	navExpect(t, "/ab", "a-star")
+	navExpect(t, "/aa", "a-star")
+	navExpect(t, "/abc", "abc")
+	navExpect(t, "/.*b", NOT_FOUND)
+	navExpect(t, "/.*c", NOT_FOUND)
+
+	navExpect(t, "/123", "123")
+
+	// the followingtest cases confirm that regex needs full string, not just subset
+	navExpect(t, "4/123", NOT_FOUND)
+	navExpect(t, "/123/123", NOT_FOUND)
+	navExpect(t, "/1234", NOT_FOUND)
+	navExpect(t, "/123/", NOT_FOUND)
+}
+
+// TestRouteWithRegexpDirs1 tests regex paths with inner wildcard
+func TestRouteWithRegexpDirs1(t *testing.T) {
+
+	resetRoutes()
+	RouteWithRegexp("/user/.*/settings", &testComp{name: "settings"})
+
+	navExpect(t, "/user/1001/settings", "settings")
+	navExpect(t, "/user/1/settings", "settings")
+	navExpect(t, "/user/1001/settings/", NOT_FOUND) // extra trailing slash
+}
+
+// TestRouteWithRegexpDirs2 test regex wildcard at end and middle
+func TestRouteWithRegexpDirs2(t *testing.T) {
+
+	resetRoutes()
+	RouteWithRegexp("/user/.*/files/.*", &testComp{name: "files"})
+
+	navExpect(t, "/user/1001/files/", "files")
+	navExpect(t, "/user/1001/files/index.html", "files")
+	navExpect(t, "/user/1001/files/foo/bar/baz.png", "files")
+	navExpect(t, "/user/team/green/files/index.html", "files")
+}
+
+// TestRouteWithRegexpDirs3 tests regex with "OR" condition
+func TestRouteWithRegexpDirs3(t *testing.T) {
+
+	resetRoutes()
+	RouteWithRegexp("/color/(red|green|blue)", &testComp{name: "color"})
+
+	navExpect(t, "/color/red", "color")
+	navExpect(t, "/color/blue", "color")
+	navExpect(t, "/color/fuschia", NOT_FOUND)
+}
+
+// TestRouteWithRegexpSharedHandler tests that a handler with internal state
+// can be re-used and shared among multiple route patterns
+func TestRouteWithRegexpHandlerWithState(t *testing.T) {
```

Comment:
I think this test does reflect the language behavior of Go related to pointers. We don't really need it.

Reply:
agreed